Skyra.Core/Cache/Models/Prompts/CoreRichDisplay.cs

```diff
@@ -0,0 +1,315 @@
+		{
+			Context = Context.Append(embed).ToArray();
+			return this;
+		}
+
+		[NotNull]
+		public CoreRichDisplay AddPage([NotNull] Func<CoreMessageEmbed, CoreMessageEmbed> callback)
+		{
+			var value = callback(Template);
+			Context = Context.Append(value).ToArray();
+			return this;
+		}
+
+		[NotNull]
+		public CoreRichDisplay SetInformationPage(CoreMessageEmbed? embed)
+		{
+			InformationPage = embed;
+			return this;
+		}
+
+		private async Task<bool> RemoveReactions(string channelId)
+		{
+			var reactionDeleteQuery =
+				IClient.Instance.Rest.Channels[channelId].Messages[MessageId.ToString()].Reactions.DeleteAsync();
+
+			return await Utilities.ResolveAsBooleanOnErrorCodes(reactionDeleteQuery, IgnoreMessageUpdateCodes);
+		}
+
+		private async Task<bool> Render(string channelId)
+		{
+			var messageUpdateQuery = IClient.Instance.Rest.Channels[channelId]
+				.Messages[MessageId.ToString()].PatchAsync(new SendableMessage
+				{
+					Embed = PagePosition == -1 ? InformationPage : Context[PagePosition]
+				});
+
+			return await Utilities.ResolveAsBooleanOnErrorCodes(messageUpdateQuery, IgnoreMessageUpdateCodes);
+		}
+
+		private CoreRichDisplayReactionType RetrieveEntry(string emoji)
+		{
+			foreach (var (type, allowedEmoji) in AllowedEmojis)
+			{
+				if (allowedEmoji == emoji) return type;
+			}
+
+			return CoreRichDisplayReactionType.None;
+		}
+
+		[ItemNotNull]
+		public async Task<CoreRichDisplay> SetUpAsync([NotNull] CoreMessage message,
+			CoreRichDisplayRunOptions? options = null)
+		{
+			options ??= new CoreRichDisplayRunOptions();
+			PagePosition = options.StartPage;
+			if (!FooterEnabled) SetFooters();
+
+			if (message.AuthorId != message.Client.Id)
+			{
+				var sent = await message.SendAsync(options.MessageContent);
+				MessageId = sent.Id;
+			}
+
+			var hasSetUp = await AddReactions(message, IterateEmojis(options.Stop, options.Jump, options.FirstLast));
+			if (!hasSetUp) return this;
+
+			await message.Client.Cache.Prompts.SetAsync(
+				new CorePromptState(message.Client, CorePromptStateType.RichDisplay, this), options.Duration);
+			await Render(message.ChannelId.ToString());
+			return this;
+		}
+
+		private async Task<bool> AddReactions([NotNull] CoreMessage message,
+			[NotNull] IEnumerable<(CoreRichDisplayReactionType, string)> emojis)
+		{
+			var reacted = false;
+			var messageId = MessageId.ToString();
+			var channelId = message.ChannelId.ToString();
+			AllowedEmojis = emojis.ToArray();
+			foreach (var (_, emoji) in AllowedEmojis)
+			{
+				var query = message.Client.Rest.Channels[channelId].Messages[messageId].Reactions[emoji]["@me"]
+					.PutAsync<object>(null);
+
+				if (!await Utilities.ResolveAsBooleanOnErrorCodes(query, IgnoreReactionRemoveCodes))
+				{
+					return false;
+				}
+
+				reacted = true;
+			}
+
+			return reacted;
+		}
+
+		private IEnumerable<(CoreRichDisplayReactionType, string)> IterateEmojis(bool stop, bool jump, bool firstLast)
```

Comment:
you can name tuples using the syntax (Type: name, Type: name) and later, when you use this method, you can deconstruct those names (instead of using .Item1, .Item2, you can use .name) - note: it's standard to pascal case said names.